include/oneapi/dpl/internal/dynamic_selection_impl/auto_tune_policy.h

```diff
@@ -0,0 +1,271 @@
+// -*- C++ -*-
+//===----------------------------------------------------------------------===//
+//
+// Copyright (C) 2023 Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef _ONEDPL_AUTO_TUNE_POLICY_H
+#define _ONEDPL_AUTO_TUNE_POLICY_H
+
+#include <stdexcept>
+#include <map>
+#include <memory>
+#include <mutex>
+#include <utility>
+#include <chrono>
+#include <limits>
+#include <vector>
+#include <type_traits>
+#include <tuple>
+#include "oneapi/dpl/internal/dynamic_selection_traits.h"
+#if _DS_BACKEND_SYCL != 0
+    #include "oneapi/dpl/internal/dynamic_selection_impl/sycl_backend.h"
+#endif
+namespace oneapi {
+namespace dpl {
+namespace experimental {
+#if _DS_BACKEND_SYCL != 0
+  template <typename Backend=sycl_backend, typename... KeyArgs>
+#else
+  template <typename Backend, typename... KeyArgs>
+#endif
+  class auto_tune_policy {
+
+    static constexpr double never_resample = 0.0;
+    static constexpr int use_best_resource = -1;
+
+    using wrapped_resource_t = typename std::decay_t<Backend>::execution_resource_t;
+    using size_type = typename std::vector<typename Backend::resource_type>::size_type;
+
+    using timing_t = uint64_t;
+
+    struct resource_with_index_t {
+      wrapped_resource_t r_;
+      size_type index_;
+    };
+
+    struct time_data_t {
+      uint64_t num_timings_ = 0;
+      timing_t value_ = 0;
+    };
+
+    struct tuner_t {
+      std::mutex m_;
+
+      std::chrono::steady_clock::time_point t0_;
+
+      timing_t best_timing_ = std::numeric_limits<timing_t>::max();
+      resource_with_index_t best_resource_;
+
+      const size_type max_resource_to_profile_;
+      uint64_t next_resource_to_profile_ = 0; // as index in resources
+
+      using time_by_index_t = std::map<size_type, time_data_t>;
+      time_by_index_t time_by_index_;
+
+      double resample_time_ = 0.0;
```

Comment:
Just to do not forget, changing a `double` to `std::uint64_t` for resampling type in all places